specs/math.txt

```diff
@@ -71,57 +82,97 @@ $\$$
-dollar can be contained: $`foo$bar`$
+these are not math texts: $ y=x$, $y=x $, $
+y=x$ and $y=x
+$
 .
-<p>this is math inline <span class="math inline">y=x</span> test.
-dollar can be contained: <span class="math inline">foo$bar</span></p>
+<p>these are not math texts: $ y=x$, $y=x $, $
+y=x$ and $y=x
+$</p>
 ````````````````````````````````
 
-Math expression without backtick cannot start/end with spaces but Math expression with backtick can:
+Inline math expressions do not need to be surrounded with whitespace:
 
 ```````````````````````````````` example
-these are not math texts: $ y=x$ and $y=x $ and $ y=x $
-but these are math texts: $` y=x`$ and $`y=x `$ and $` y=x `$
+these are math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
 .
-<p>these are not math texts: $ y=x$ and $y=x $ and $ y=x $
-but these are math texts: <span class="math inline"> y=x</span> and <span class="math inline">y=x </span> and <span class="math inline"> y=x </span></p>
+<p>these are math texts: foo<span class="math inline">y=x</span>bar and <span class="math inline">y=x</span>bar and foo<span class="math inline">y=x</span> bar</p>
 ````````````````````````````````
 
-Math expression without backtick surrounded with whitespaces:
+Math expressions can be immediately followed by other math expressions:
 
 ```````````````````````````````` example
-these are not math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
-but these are math texts: foo$`y=x`$bar and $`y=x`$bar and foo$`y=x`$ bar
+$a$$b$
+
+$a$$$b$$
+
+$$a$$$b$
+
+$$a$$$$b$$
 .
-<p>these are not math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
-but these are math texts: foo<span class="math inline">y=x</span>bar and <span class="math inline">y=x</span>bar and foo<span class="math inline">y=x</span> bar</p>
+<p><span class="math inline">a</span><span class="math inline">b</span></p>
+<p><span class="math inline">a</span><span class="math display">b</span></p>
+<p><span class="math display">a</span><span class="math inline">b</span></p>
+<p><span class="math display">a</span><span class="math display">b</span></p>
 ````````````````````````````````
 
-Math expression without backtick can be end with punctuation but cannot start with:
+Both inline and display mode math expressions are inline elements with the same
+precedence as code spans. The leftmost valid element takes priority:
 
 ```````````````````````````````` example
-math texts: $x=y$! and $x=y$? and $x=y$: and $x=y$.
-not math texts: !$x=y$! and ?x=y$? and :$x=y$: and .$x=y$.
-backticks doesn't have the edge case: !$`x=y`$! and ?`x=y`$? and :$`x=y`$: and .$`x=y`$.
+$Inline `first$ then` code
+
+`Code $first` then$ inline
+
+$$ Display `first $$ then` code
+
+`Code $$ first` then $$ display
 .
-<p>math texts: <span class="math inline">x=y</span>! and <span class="math inline">x=y</span>? and <span class="math inline">x=y</span>: and <span class="math inline">x=y</span>.
-not math texts: !$x=y$! and ?x=y$? and :$x=y$: and .$x=y$.
```

Comment:
Can you keep test cases where some non-alphabetical characters are just before/after inline math expressions? (e.g. `This is math expression $y=x$!`)